h2o-core/src/main/java/water/parser/Parser.java

```diff
@@ -83,30 +89,120 @@ ParseWriter streamParseZip( final InputStream is, final StreamParseWriter dout,
+    assert zidx == 1;
+    boolean goodFile = true;       // whether to sparse a file or not.  Some zip directory may contain
+                                    // junk info or dataset from other bigger datasets by mistake
+//    int count = 0;
+
+    while (is.available() > 0) {
+      int xidx = bvs.read(null, 0, 0); // Back-channel read of chunk index
+      if (xidx > zidx) {  // Advanced chunk index of underlying ByteVec stream?
         zidx = xidx;       // Record advancing of chunk
         nextChunk.close(); // Match output chunks to input zipfile chunks
-        if( dout != nextChunk ) {
+        if (dout != nextChunk) {
           dout.reduce(nextChunk);
           if (_jobKey != null && _jobKey.get().stop_requested()) break;
         }
         nextChunk = nextChunk.nextChunk();
       }
-      parseChunk(cidx++, din, nextChunk);
+
+      if (cidx == 0) {
+        // We perform guess setup again and make sure that the files contain parts of the same dataset.  By that,
+        // we mean same number of columns, same column types.  This check is performed to make sure users may have
+        // added one extra file by mistake.  Instead of throwing an error, we just won't parse it.  Will generate
+        // a warning to user about this.  However, we do assume that you zip the same file types inside a directory.
+        // This is the best Tomas and I have come up with.  It is not perfect but
+        // it should work.  In addition, we still require that only the first file in the zip directory has header
+        // in them if the user wants to add the header names inside the dataset.  This is required so that flow and
+        // python/R clients respond the same way.  Again, this is not perfect.
+        byte[] headerBytes;
+
+        try { // reading system file will cause a null pointer exception here.
+          headerBytes = ZipUtil.unzipForHeader(din.getChunkData(cidx), this._setup._chunk_size);
+          ParseSetup ps = ParseSetup.guessSetup(null, headerBytes, GUESS_INFO, ParseSetup.GUESS_SEP,
+                  ParseSetup.GUESS_COL_CNT, this._setup._single_quotes, ParseSetup.GUESS_HEADER,
+                  null, null, null, null);
+
+          // check to make sure datasets in file belong to the same dataset
+          // just check for number for number of columns here.  Ignore the column type, user can force it
+          if (this._setup._number_columns == ps._number_columns)
+            goodFile = true;
+          else {
+            String warning = "Your zip file contains a file that belong to another dataset with different " +
+                    "number of column.  Number of columns for files that have been parsed = "+
+                    this._setup._number_columns + ".  Number of columns in new file = "+ps._number_columns+
+                    ".  This new file is skipped and not parsed.";
+            dout.addError(new ParseWriter.ParseErr(warning, -1, -1L, -2L));
+            goodFile = false;
```

Comment:
I think we should skip to the next file instead and hit continue, otherwise it's confusing